cmd/s3-layer-anonymous.go

```diff
@@ -3,32 +3,62 @@ package cmd
 import "io"
 
 // AnonGetObject - Get object anonymously
-func (l *S3Layer) AnonGetObject(bucket string, key string, startOffset int64, length int64, writer io.Writer) (err error) {
-	// already an s3 object error
-	return l.GetObject(bucket, key, startOffset, length, writer)
+func (l *s3Gateway) AnonGetObject(bucket string, key string, startOffset int64, length int64, writer io.Writer) error {
+	object, err := l.anonClient.GetObject(bucket, key)
+	if err != nil {
+		return s3ToObjectError(traceError(err), bucket, key)
+	}
+
+	object.Seek(startOffset, io.SeekStart)
+	if _, err := io.CopyN(writer, object, length); err != nil {
+		return s3ToObjectError(traceError(err), bucket, key)
+	}
+
+	object.Close()
+	return nil
 }
 
 // AnonGetObjectInfo - Get object info anonymously
-func (l *S3Layer) AnonGetObjectInfo(bucket string, object string) (objInfo ObjectInfo, err error) {
-	// already an s3 object error
-	return l.GetObjectInfo(bucket, object)
+func (l *s3Gateway) AnonGetObjectInfo(bucket string, object string) (ObjectInfo, error) {
+	oi, err := l.anonClient.StatObject(bucket, object)
+	if err != nil {
+		return ObjectInfo{}, s3ToObjectError(traceError(err), bucket, object)
+	}
+
+	return fromMinioClientObjectInfo(bucket, oi), nil
 }
 
 // AnonListObjects - List objects anonymously
-func (l *S3Layer) AnonListObjects(bucket string, prefix string, marker string, delimiter string, maxKeys int) (result ListObjectsInfo, err error) {
-	result, err = l.ListObjects(bucket, prefix, marker, delimiter, maxKeys)
+func (l *s3Gateway) AnonListObjects(bucket string, prefix string, marker string, delimiter string, maxKeys int) (ListObjectsInfo, error) {
+	loi := ListObjectsInfo{}
+
+	loi.Objects = make([]ObjectInfo, maxKeys)
+
+	result, err := l.anonClient.ListObjects(bucket, prefix, marker, delimiter, maxKeys)
 	if err != nil {
-		return ListObjectsInfo{}, s3ToObjectError(traceError(err), bucket)
+		return ListObjectsInfo{}, err
 	}
-	return
+
+	return fromMinioClientListBucketResult(bucket, result), nil
 }
 
 // AnonGetBucketInfo - Get bucket metadata anonymously.
-func (l *S3Layer) AnonGetBucketInfo(bucket string) (bucketInfo BucketInfo, err error) {
-	bucketInfo, err = l.GetBucketInfo(bucket)
+func (l *s3Gateway) AnonGetBucketInfo(bucket string) (BucketInfo, error) {
```

Comment:
Since S3 does not allow anonymous listing of buckets, what you can do is do `l.anonClient.BucketExists()` if there is no error then do authenticated ListBuckets to get the `creationDate`, if there was any error from `l.anonClient.BucketExists()` then we can just return error.